Source/Diagnostics/WarpXOpenPMD.H

```diff
@@ -0,0 +1,126 @@
+      amrex::ParallelDescriptor::ReduceRealMax(m_End, amrex::ParallelDescriptor::IOProcessorNumber());
+      amrex::Print()<<m_Tag<<" took:"<<m_End - m_Start<<" seconds\n";
+  }
+private:
+  amrex::Real m_Start;
+  amrex::Real m_End;
+  std::string m_Tag;
+};
+
+
+//
+//
+class WarpXParticleCounter
+{
+public:
+  WarpXParticleCounter(const std::unique_ptr<WarpXParticleContainer>& pc);
+  unsigned long GetTotalNumParticles() {return m_Total;}
+
+  std::vector<unsigned long long> m_ParticleOffsetAtRank;;
+  std::vector<unsigned long long> m_ParticleSizeAtRank;;
+private:
+  // get the offset in the overall particle id collection
+  void GetParticleOffsetOfProcessor(const long& numParticles,
+                    unsigned long long& offset,
+                    unsigned long long& sum)  const ;
+
+
+  int m_MPIRank = 0;
+  int m_MPISize = 1;
+
+  unsigned long long m_Total = 0;
+
+  std::vector<unsigned long long> m_ParticleCounterByLevel;
+};
+
+
+//
+//
+//
+class WarpXOpenPMDPlot
+{
+public:
+  WarpXOpenPMDPlot(bool, const std::string& filetype);
+  //WarpXOpenPMDPlot(const std::string& dir, const std::string& fileType);
+  ~WarpXOpenPMDPlot();
+
+  void SetStep(int ts);
+  void SetOpenPMDType(const std::string& type);
+
+  void WriteOpenPMDParticles(const std::unique_ptr<MultiParticleContainer>&);
+
+  void WriteOpenPMDFields(//const std::string& filename,
+              const std::vector<std::string>& varnames,
+              const amrex::MultiFab& mf,
+              const amrex::Geometry& geom,
+              const int iteration, const double time ) const;
+
+
+private:
+  void Init(const std::string& filename,
+        openPMD::AccessType accessType);
+
+  /*
+  // get the offset in the overall particle id collection
+  void GetParticleOffsetOfProcessor(const long& numParticles,
+                    unsigned long long& offset,
+                    unsigned long long& sum)  const ;
+  */
+  void SetupPos(openPMD::ParticleSpecies& currSpecies,
+        const unsigned long long& np) const ;
+
+  void SetupRealProperties(openPMD::ParticleSpecies& currSpecies,
+               const amrex::Vector<int>& write_real_comp,
+               const amrex::Vector<std::string>& real_comp_names,
+               unsigned long long np) const;
+
+  void SaveRealProperty(WarpXParIter& pti, //int, int,
+            openPMD::ParticleSpecies& currSpecies,
+            unsigned long long offset,
+            const amrex::Vector<int>& write_real_comp,
+            const amrex::Vector<std::string>& real_comp_names) const;
+
+
+  void SavePlotFile(const std::unique_ptr<WarpXParticleContainer>& pc,
+            const std::string& name,
+            int iteration,
+            const amrex::Vector<int>& write_real_comp,
+            const amrex::Vector<int>& write_int_comp,
+            const amrex::Vector<std::string>& real_comp_names,
+            const amrex::Vector<std::string>&  int_comp_names) const;
+
+
+  void GetFileName(std::string& filename, int ts);
```

Comment:
openPMD-api supports a regex for the time step (e.g. `%07T`) in the file name, so we do not need to pass this here:
```suggestion
  void GetFileName(std::string& filename);
```
Refs.:
- https://github.com/openPMD/openPMD-standard/blob/1.1.0/STANDARD.md#iterations-and-time-series
- https://www.openpmd.org/openPMD-api/classopen_p_m_d_1_1_series.html#afe9b974cd85678c4df076a18e46fcfe3

Sorry for the missing doc-string in `Series::Series` about this, I added a comment with a code example in the implementation file to show how this can be used :-)